File: automations/seo-optimizer/02-agents/_shared/slack_blockkit.py
"""Slack Block Kit message builders.

The outbox stores Slack payloads as JSON in `notifications_outbox.payload`.
This module builds those payloads. Workers read them and POST to Slack.

We build two types of messages:
  - SEO notification: "Run completed — N opportunities for review"
  - Writer notification: "Article rewrite ready for [Client] / [Article title]"
  - Admin alert: pipeline failures, stuck states (from watchdog)
"""
from __future__ import annotations


def build_seo_notification(
    *,
    client_name: str,
    run_id: str,
    opportunities_count: int,
    by_category: dict[str, int],
    review_url: str,
) -> dict:
    """Notification to SEO specialist when a new run completes."""
    cat_summary = " · ".join(f"{cat}: {n}" for cat, n in by_category.items())
    return {
        "text": f"🎯 {opportunities_count} oportunidades SEO para {client_name}",
        "blocks": [
            {
                "type": "header",
                "text": {"type": "plain_text", "text": f"🎯 {opportunities_count} oportunidades SEO — {client_name}"},
            },
            {
                "type": "section",
                "text": {"type": "mrkdwn", "text": f"*Por categoría:* {cat_summary}"},
            },
            {
                "type": "actions",
                "elements": [
                    {
                        "type": "button",
                        "text": {"type": "plain_text", "text": "Revisar en Orbit"},
                        "url": review_url,
                        "style": "primary",
                    },
                ],
            },
            {
                "type": "context",
                "elements": [
                    {"type": "mrkdwn", "text": f"Run `{run_id}` · seo-optimizer"},
                ],
            },
        ],
    }


def build_writer_notification(
    *,
    client_name: str,
    article_title: str,
    article_url: str,
    category: str,
    change_summary: str,
    inbox_url: str,
) -> dict:
    """Notification to redactor when a rewrite is ready."""
    return {
        "text": f"✍️ Reescritura lista — {article_title}",
        "blocks": [
            {
                "type": "header",
                "text": {"type": "plain_text", "text": f"✍️ Reescritura lista — {client_name}"},
            },
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": (
                        f"*Artículo:* <{article_url}|{article_title}>\n"
                        f"*Categoría:* `{category}`\n"
                        f"*Resumen del cambio:*\n{change_summary}"
                    ),
                },
            },
            {
                "type": "actions",
                "elements": [
                    {
                        "type": "button",
                        "text": {"type": "plain_text", "text": "Abrir en Orbit"},
                        "url": inbox_url,
                        "style": "primary",
                    },
                ],
            },
        ],
    }


def build_admin_alert(
    *,
    severity: str,        # 'RED' | 'YELLOW'
    title: str,
    body: str,
    details: dict | None = None,
) -> dict:
    """Watchdog / failure alert to admin channel."""
    emoji = "🔴" if severity == "RED" else "🟡"
    fields = []
    if details:
        for k, v in details.items():
            fields.append({"type": "mrkdwn", "text": f"*{k}:* `{v}`"})
    blocks: list[dict] = [
        {
            "type": "header",
            "text": {"type": "plain_text", "text": f"{emoji} {title}"},
        },
        {
            "type": "section",
            "text": {"type": "mrkdwn", "text": body},
        },
    ]
    if fields:
        blocks.append({"type": "section", "fields": fields[:10]})  # Slack max 10
    blocks.append({
        "type": "context",
        "elements": [{"type": "mrkdwn", "text": "seo-optimizer · watchdog"}],
    })
    return {"text": f"{emoji} {title}", "blocks": blocks}


def build_reeval_outcome_alert(
    *,
    client_name: str,
    article_title: str,
    article_url: str,
    outcome: str,         # 'improved' | 'worsened' | 'unchanged' | 'inconclusive'
    metrics: dict,
) -> dict:
    """Alert to SEO when a re-eval lands — especially important for 'worsened'."""
    emoji = {
        "improved": "✅",
        "worsened": "⚠️",
        "unchanged": "➖",
        "inconclusive": "❓",
    }.get(outcome, "❓")
    return {
        "text": f"{emoji} Re-evaluación: {article_title}",
        "blocks": [
            {
                "type": "header",
                "text": {"type": "plain_text", "text": f"{emoji} Re-evaluación a 45 días — {client_name}"},
            },
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": (
                        f"*Artículo:* <{article_url}|{article_title}>\n"
                        f"*Resultado:* `{outcome}`\n"
                        f"*Clicks:* {metrics.get('clicks_before',0)} → {metrics.get('clicks_after',0)} "
                        f"(Δ {metrics.get('clicks_delta_pct',0):+.1f}%)\n"
                        f"*Posición:* {metrics.get('position_before',0):.1f} → {metrics.get('position_after',0):.1f} "
                        f"(Δ {metrics.get('position_delta',0):+.1f})"
                    ),
                },
            },
        ],
    }

File: automations/seo-optimizer/supabase/functions/_shared/slack-blockkit.ts
// _shared/slack-blockkit.ts
// Builders de Slack Block Kit. Solo construyen el JSON; el envío lo hace outbox-worker.

export interface SeoNotificationArgs {
  clientName: string;
  runId: string;
  opportunitiesCount: number;
  byCategory: Record<string, number>;
  reviewUrl: string;
}

export function buildSeoNotification(a: SeoNotificationArgs): Record<string, unknown> {
  const catSummary = Object.entries(a.byCategory)
    .map(([cat, n]) => `${cat}: ${n}`)
    .join(" · ");
  return {
    text: `🎯 ${a.opportunitiesCount} oportunidades SEO para ${a.clientName}`,
    blocks: [
      {
        type: "header",
        text: { type: "plain_text", text: `🎯 ${a.opportunitiesCount} oportunidades SEO — ${a.clientName}` },
      },
      {
        type: "section",
        text: { type: "mrkdwn", text: `*Por categoría:* ${catSummary}` },
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: { type: "plain_text", text: "Revisar en Orbit" },
            url: a.reviewUrl,
            style: "primary",
          },
        ],
      },
      {
        type: "context",
        elements: [{ type: "mrkdwn", text: `Run \`${a.runId}\` · seo-optimizer` }],
      },
    ],
  };
}

export interface WriterNotificationArgs {
  clientName: string;
  articleTitle: string;
  articleUrl: string;
  category: string;
  changeSummary: string;
  inboxUrl: string;
}

export function buildWriterNotification(a: WriterNotificationArgs): Record<string, unknown> {
  return {
    text: `✍️ Reescritura lista — ${a.articleTitle}`,
    blocks: [
      {
        type: "header",
        text: { type: "plain_text", text: `✍️ Reescritura lista — ${a.clientName}` },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text:
            `*Artículo:* <${a.articleUrl}|${a.articleTitle}>\n` +
            `*Categoría:* \`${a.category}\`\n` +
            `*Resumen del cambio:*\n${a.changeSummary}`,
        },
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: { type: "plain_text", text: "Abrir en Orbit" },
            url: a.inboxUrl,
            style: "primary",
          },
        ],
      },
    ],
  };
}

export interface AdminAlertArgs {
  severity: "RED" | "YELLOW";
  title: string;
  body: string;
  details?: Record<string, string | number>;
}

export function buildAdminAlert(a: AdminAlertArgs): Record<string, unknown> {
  const emoji = a.severity === "RED" ? "🔴" : "🟡";
  const fields: Array<{ type: string; text: string }> = [];
  if (a.details) {
    for (const [k, v] of Object.entries(a.details)) {
      fields.push({ type: "mrkdwn", text: `*${k}:* \`${v}\`` });
    }
  }
  const blocks: Array<Record<string, unknown>> = [
    { type: "header", text: { type: "plain_text", text: `${emoji} ${a.title}` } },
    { type: "section", text: { type: "mrkdwn", text: a.body } },
  ];
  if (fields.length > 0) {
    blocks.push({ type: "section", fields: fields.slice(0, 10) });
  }
  blocks.push({
    type: "context",
    elements: [{ type: "mrkdwn", text: "seo-optimizer · watchdog" }],
  });
  return { text: `${emoji} ${a.title}`, blocks };
}

export interface ReevalOutcomeAlertArgs {
  clientName: string;
  articleTitle: string;
  articleUrl: string;
  outcome: "improved" | "worsened" | "unchanged" | "inconclusive";
  metrics: {
    clicks_before?: number; clicks_after?: number; clicks_delta_pct?: number;
    position_before?: number; position_after?: number; position_delta?: number;
  };
}

export function buildReevalOutcomeAlert(a: ReevalOutcomeAlertArgs): Record<string, unknown> {
  const emojiMap: Record<string, string> = { improved: "✅", worsened: "⚠️", unchanged: "➖", inconclusive: "❓" };
  const emoji = emojiMap[a.outcome] ?? "❓";
  const m = a.metrics;
  return {
    text: `${emoji} Re-evaluación: ${a.articleTitle}`,
    blocks: [
      { type: "header", text: { type: "plain_text", text: `${emoji} Re-evaluación a 45 días — ${a.clientName}` } },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text:
            `*Artículo:* <${a.articleUrl}|${a.articleTitle}>\n` +
            `*Resultado:* \`${a.outcome}\`\n` +
            `*Clicks:* ${m.clicks_before ?? 0} → ${m.clicks_after ?? 0} ` +
            `(Δ ${(m.clicks_delta_pct ?? 0).toFixed(1)}%)\n` +
            `*Posición:* ${(m.position_before ?? 0).toFixed(1)} → ${(m.position_after ?? 0).toFixed(1)} ` +
            `(Δ ${(m.position_delta ?? 0).toFixed(1)})`,
        },
      },
    ],
  };
}
